docs/design/repair-with-hashes.md

```diff
@@ -0,0 +1,63 @@
+> To solve this problem, hashes of every piece will be stored alongside each piece on each storage node. A validation hash that the set of hashes is correct will be stored in the pointer. During repair, the hashes of every piece can be retrieved and validated for correctness against the pointer, thus allowing each piece to be validated in its entirety. This allows the repair system to correctly assess whether or not repair has been completed successfully without using extra redundancy for the same task.
+
+Hash verification on the satellite requires understanding the current piece signing and verification workflow:
+
+1. Satellite generates a new key-pair for each piece.
+2. The piece public key is included in the order limit and signed by the satellite
+3. The piece private key is included in the create segment, download segment or delete a segment.
+4. The uplink uses piece private key to sign the orders.
+5. The storage node verifies order limit signature as usual and then uses the piece public key to verify the order and piece hash.
+6. Storage node sends both the order limit and order to the satellite during settlement.
+7. Storage node sends data, order limit and piece hash to the satellite, if it needs to prove data-consistency to the satellite.
+
+Thus the uplink-signed `uplink_piece_hash` is already stored in the storage node's `pieceinfo_` table as a serialized protocol buffer and includes the `PieceID`, `Hash`, and uplink `Signature`. The `pieceinfo_` table also includes the `order_limit` which contains the `PiecePublicKey` necessary for validating the `PieceHash`, as well as the `SatelliteSignature` to validate the `OrderLimit` itself.
+
+## Design
+
+When a satellite is download pieces for repair, the storage node should also include the `PieceHash` and `OrderLimit` to the satellite along with each piece.
+
+The satellite should attempt to download exactly the Reed-Solomon minimum number of pieces during a repair, storing them in memory. It should validate the `OrderLimit` using its own private key.  It should validate that the `PieceHash` `Signature` corresponds to the `PublicKey` contained in the `OrderLimit`.  The satellite should calculate the size and hash of the piece as it downloads them, and it should be equal to the `PieceHash` `Hash` and `PieceSize`. If a piece fails any of these checks, an error should be logged and a different piece should be downloaded.  Once all pieces are downloaded, the `PiecePublicKey` of all pieces should be compared and equal.  If they are not equal, an error should be logged and more pieces should be downloaded until a minimum number of pieces share the same `PiecePublicKey`.  When enough pieces have been downloaded and validated, Reed-Solomon decoding should be used to reassemble the segment, and repair should continue as previously designed.
```

Comment:
just because there's a lot of vocab and things moving around here, it could be helpful to include what a struct would look like for a `PiecePublicKey` or even `OrderLimit` here even though that isn't new.